src/ceph-volume/ceph_volume/util/disk.py

```diff
@@ -221,3 +228,205 @@ def is_partition(dev):
+    suffix_index = 0
+    while size > 1024:
+        suffix_index += 1
+        size = size / 1024.0
+    return "{size:.2f} {suffix}".format(
+        size=size,
+        suffix=suffixes[suffix_index])
+
+
+def get_partitions_facts(sys_block_path):
+    partition_metadata = {}
+    for folder in os.listdir(sys_block_path):
+        folder_path = os.path.join(sys_block_path, folder)
+        if os.path.exists(os.path.join(folder_path, 'partition')):
+            contents = get_file_contents(os.path.join(folder_path, 'partition'))
+            if '1' in contents:
+                part = {}
+                partname = folder
+                part_sys_block_path = os.path.join(sys_block_path, partname)
+
+                part['start'] = get_file_contents(part_sys_block_path + "/start", 0)
+                part['sectors'] = get_file_contents(part_sys_block_path + "/size", 0)
+
+                part['sectorsize'] = get_file_contents(
+                    part_sys_block_path + "/queue/logical_block_size")
+                if not part['sectorsize']:
+                    part['sectorsize'] = get_file_contents(
+                        part_sys_block_path + "/queue/hw_sector_size", 512)
+                part['size'] = human_readable_size(float(part['sectors']) * 512)
+
+                partition_metadata[partname] = part
+    return partition_metadata
+
+
+def get_devices(_sys_block_path='/sys/block', _dev_path='/dev', _mapper_path='/dev/mapper'):
+    """
+    Captures all available devices from /sys/block/, including its partitions,
+    along with interesting metadata like sectors, size, vendor,
+    solid/rotational, etc...
+
+    Returns a dictionary, where keys are the full paths to devices.
+
+    ..note:: dmapper devices get their path updated to what they link from, if
+            /dev/dm-0 is linked by /dev/mapper/ceph-data, then the latter gets
+            used as the key.
+
+    ..note:: loop devices, removable media, and logical volumes are never included.
+    """
+    # Portions of this detection process are inspired by some of the fact
+    # gathering done by Ansible in module_utils/facts/hardware/linux.py. The
+    # processing of metadata and final outcome *is very different* and fully
+    # imcompatible. There are ignored devices, and paths get resolved depending
+    # on dm devices, loop, and removable media
+
+    device_facts = {}
+
+    block_devs = get_block_devs(_sys_block_path)
+    dev_devs = get_dev_devs(_dev_path)
+    mapper_devs = get_mapper_devs(_mapper_path)
+
+    for block in block_devs:
+        sysdir = os.path.join(_sys_block_path, block)
+        metadata = {}
+
+        # Ensure that the diskname is an absolute path and that it never points
+        # to a /dev/dm-* device
+        diskname = mapper_devs.get(block) or dev_devs.get(block)
+
+        # If the mapper device is a logical volume it gets excluded
+        if diskname.startswith(('/dev/mapper', '/dev/dm-')):
```

Comment:
I'm not a big fan of rejecting devices during the detection.
The detection code will be large and needs to be splitted in another function.
This code should only take care of detecting the devices and another to filter out the invalid devices. That would avoid the implicit rejection of loop devices and a long & intricated list in this code.

Reply:
By the way, this code is lacking of rejecting the following kind of devices:
- read-only devices
- in-use partitions
- in-use raw devices
- already configured disks with ceph-disk

Reply:
We are explicitly not allowing partitions so no need to detect partitions, we do detect raw devices that might be in use by LVM, and we don't support ceph-disk since it is deprecated, and it doesn't allow raw devices or LVs to be consumed

Reply:
So how do you handle existing configurations ? We cannot ignore them.

Reply:
ceph-volume at this point in time can certainly ignore them. Existing configurations will have to manually zap their devices to get rid of partitions and ensure their raw partitions are not in use by LVM. For more advanced usage, ceph-ansible would be a better choice no doubt, which is why I am suggesting that you should get your functionality in ceph-ansible as well

Reply:
I certainly missed something here. I cannot understand how ceph-volume could ignore all the existing running Ceph instances when it tries to detect free devices.